pkg/strace/strace.go

```diff
@@ -0,0 +1,165 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package strace supports tracing programs.
+// The basic control of tracing is via a Tracer, which returns raw
+// TraceRecords via a chan. The easiest way to crate a Tracer is via
+// NewTraceCommand, which creates a TraceCmd which embeds exec.Cmd and
+// implements parts that add tracing.
+// Output from tracing defaults to os.Stderr, but can be changed by a
+// call to SetOutput, which accepts a WriteCloser.
+// The intent is that one setups up a TraceCmd, calls Start, and then
+// consumes the records.
+// Forking and nice printing are not yet supported.
+package strace
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+var Debug = func(string, ...interface{}) {}
+
+// TraceCmd implements the parts of exec.Cmd needed to
+// create and run a traced process
+type TraceCmd struct {
+	*exec.Cmd
+	out io.WriteCloser
+}
+
+// NewTraceCommand sets up a basic exec.Cmd
+func NewTraceCommand(n string, a ...string) (*TraceCmd, error) {
+	t := &TraceCmd{Cmd: exec.Command(n, a...), out: os.Stderr}
+	return t, nil
+}
+
+// EventType describes whether a record is system call Entry or Exit
+type EventType string
+
+const (
+	Enter EventType = "E"
+	Exit            = "X"
+)
+
+// TraceRecord has information about a ptrace event.
+type TraceRecord struct {
+	EX     EventType
+	Regs   unix.PtraceRegs
+	Serial int
+	Pid    int
+	Err    error
+}
+
+// String is a stringer for TraceRecords
+// TODO: stringer for Regs.
+func (t *TraceRecord) String() string {
+	pre := fmt.Sprintf("%s %d#%d:", t.EX, t.Pid, t.Serial)
+	if t.Err != nil {
+		return fmt.Sprintf("%s(%v)", pre, t.Err)
+	}
+	return fmt.Sprintf("%s %v", pre, t.Regs)
+}
+
+// Tracer has information to trace one process. It can be created by
+// starting a command, or attaching. Attaching is not supported yet.
+type Tracer struct {
+	Pid     int
+	Records chan *TraceRecord
+	Count   int
+}
+
+// Start starts a TraceCmd and returns a Tracer and error.
+func (c *TraceCmd) Start() (*Tracer, error) {
+	if c.Cmd.SysProcAttr == nil {
+		c.Cmd.SysProcAttr = &syscall.SysProcAttr{}
+	}
+	c.Cmd.SysProcAttr.Ptrace = true
+	if err := c.Cmd.Start(); err != nil {
+		Debug("Start gets err %v", err)
+		return nil, err
+	}
+	Debug("Start gets pid %v", c.Process.Pid)
+	return NewTracer(c.Process.Pid)
+}
+
+// SetOutput accepts an output to be used by Run(). If no output
+// is set Run() will use the default set in NewTraceCommand
+func (c *TraceCmd) SetOutput(w io.WriteCloser) {
+	c.out = w
+}
+
+// Run will Start the Cmd to create a Tracer, start the Tracer,
+// and then read each trace record and write it to the c.out.
+// It returns only on error or when the process exits.
+func (c *TraceCmd) Run() error {
+	t, err := c.Start()
+	if err != nil {
+		return err
+	}
+	t.Start()
+	for r := range t.Records {
+		fmt.Fprintf(c.out, "%v\n", r)
+	}
+	c.out.Close()
```

Comment:
I feel we shouldn't close `os.Stderr` because there may be more error messages we want to print to it. Rather, would it be better for the caller to close the `io.Writer`?

Reply:
I agree. I think that there should not be a default out maybe. Let's talk a bit more, what would you recommend?

Reply:
Maybe you can make an `io.Writer` a mandatory field for `NewTraceCommand`, so the signature becomes:

    func NewTraceCommand(w io.Writer, n string, a ...string)